Fixed #34110 -- Added in-memory file storage.

panicofr opened the pull request:
Added new storage type InMemoryStorage, that keeps file in memory. It implements Storage interface and can be used as storage for FileField

pauloxnet commented on django/core/files/storage.py:
```diff
@@ -415,6 +423,220 @@ def get_modified_time(self, name):
+
+
+@deconstructible()
+class InMemoryStorage(Storage, StorageSettingsMixin):
```
Could it be useful to have each Storage implementation in a separate file? So how does it happen with the different Cache backends?

panicofr replied:
Indeed... It was actually my first implementation, then I reverted to a single-file module to keep tests about deconstruction working. I didn't check `@deconstructible` docs, my fault 😅

felixxm commented on django/core/files/storage/__init__.py:
```diff
@@ -0,0 +1,28 @@
+from django.conf import settings
```
Adding the `storage` module should be made in advance in a separate PR. We don't want to mix new features with cleanups.

panicofr replied:
Yes, good point. Should I create a ticket about this topic myself?

felixxm replied:
There is no need for a separate ticket, you can open PR titled `Refs #34110 -- ...`.

panicofr replied:
Ok thanks 👍

felixxm commented on django/core/files/storage/base.py:
```diff
@@ -0,0 +1,192 @@
+from django.utils.crypto import get_random_string
+from django.utils.text import get_valid_filename
+
+__all__ = ("Storage",)
```
Is this needed? :thinking:

felixxm reviewed (commented):
@panicofr Thanks for this patch :+1: Documentation changes are missing, we should at least:
- add release notes,
- [document `InMemoryStorage`](https://docs.djangoproject.com/en/dev/ref/files/storage/),
- add `InMemoryStorage` to the ["Speeding up the tests"](https://docs.djangoproject.com/en/dev/topics/testing/overview/#speeding-up-the-tests)" section.

felixxm commented:
I rebased after #16282.

panicofr commented:
Hi @felixxm,

does this PR need further improvements?

Sorry but I'm not acquainted with Django release cycle yet 😅 

Thanks!

felixxm commented:
@panicofr The next step is to uncheck _"Needs ..."_ flags on the ticket, this will put it back in the [review queue](https://code.djangoproject.com/query?status=!closed&needs_better_patch=0&needs_tests=0&needs_docs=0&has_patch=1&stage=Accepted&desc=1&order=changetime).

felixxm commented on docs/topics/testing/overview.txt:
```diff
@@ -366,3 +366,10 @@ Preserving the test database
 runs. It skips the create and destroy actions which can greatly decrease the
 time to run tests.
+
+Avoiding disk-access for media files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
```suggestion
Avoiding disk-access for media files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. versionadded:: 4.2
```

felixxm commented on tests/file_storage/tests.py:
```diff
@@ -1157,3 +1160,184 @@ def test_urllib_request_urlopen(self):
+        self.assertEqual(self.storage.open("file.dat", "rb").read(), b"test")
+
+    def test_file_size(self):
+        """
+        Check the size of in-memory files
+        """
```
To add more explanation, docstrings should state the expected behavior and omit prefixes like "Check" since all tests check things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

carltongibson commented on docs/ref/files/storage.txt:
```diff
@@ -74,6 +74,22 @@ The ``FileSystemStorage`` class
+
+    The :class:`~django.core.files.storage.InMemoryStorage` class implements
+    a memory-based file storage. It has no persistence, but can be useful for
+    speeding tests up avoiding disk access.
```
```suggestion
    speeding tests up by avoiding disk access.
```

carltongibson commented on docs/releases/4.2.txt:
```diff
@@ -26,6 +26,12 @@ and only officially support the latest release of each series.
+In-memory files storage
+-----------------------
+
+The new ``django.core.files.storage.InMemoryStorage`` class provides a new
+non-persistent storage type useful for speeding tests up.
```
Drop the 2nd `new` perhaps.

felixxm commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,231 @@
+        self._content_type = type(content)
+        self._initialize_stream()
+
+        self.created_time = now()
+        self.accessed_time = now()
+        self.modified_time = now()
```
Should we call `now()` three times?
```suggestion
        self.created_time = now()
        self.accessed_time = self.created_time
        self.modified_time = self.created_time
```

As far as I'm aware all dates should be the same when a file is created.

felixxm commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,231 @@
+
+
+class InMemoryFileNode(ContentFile):
+    """
+    Helper class representing an in-memory file node
+    """
```
I don't see much value in this docstring.

felixxm commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,231 @@
+
+    def _initialize_stream(self):
+        """
+        Initialize underlying stream according to content type
```
```suggestion
        Initialize underlying stream according to the content type.
```

felixxm commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,231 @@
+
+    def _convert_stream_content(self, mode):
+        """
+        Convert actual file content according to opening mode
```
```suggestion
        Convert actual file content according to the opening mode.
```

felixxm commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,231 @@
+
+    def _convert_content(self, content):
+        """
+        Convert string to bytes and back, according to current content type
```
```suggestion
        Convert string to bytes and back, according to the current content
        type.
```

felixxm commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,231 @@
+
+
+class InMemoryDirNode:
+    """
+    Helper class representing an in-memory directory node
+    """
```
I don't see much value in this docstring.

felixxm commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,231 @@
+
+            # If current node is a file node and there are unprocessed segments,
+            # raise an error
+            if isinstance(current_node, InMemoryFileNode):
+                path_segments = path.split("/")
+                current_path = "/".join(
+                    path_segments[: path_segments.index(path_segment)]
+                )
+                raise NotADirectoryError(
+                    errno.ENOTDIR, os.strerror(errno.ENOTDIR), current_path
+                )
```
This is untested.

felixxm commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,231 @@
+            url = url.lstrip("/")
+        return urljoin(self.base_url, url)
+
+    def get_accessed_time(self, name):
+        file_node = self._root.resolve(name, leaf_cls=InMemoryFileNode)
+        return file_node.accessed_time
+
+    def get_created_time(self, name):
+        file_node = self._root.resolve(name, leaf_cls=InMemoryFileNode)
+        return file_node.created_time
+
+    def get_modified_time(self, name):
+        file_node = self._root.resolve(name, leaf_cls=InMemoryFileNode)
+        return file_node.modified_time
```
Why `InMemoryDirNode` doesn't have times?

felixxm commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,231 @@
+    def delete(self, name):
+        path, filename = os.path.split(name)
+        dir_node = self._root.resolve(path)
+        if dir_node is None:
+            return None
```
This is untested.

felixxm replied:
This is still untested.

felixxm reviewed (commented):
@panicofr Thanks :+1: I left initial comments.

pauloxnet commented:
@panicofr thanks for your work here I would find the time review this again.

pauloxnet commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,243 @@
+        fd = None
+
+        for chunk in content.chunks():
+            if fd is None:
```
Coverage says:
"line 200 didn't jump to line 203, because the condition on line 200 was never false"

pauloxnet commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,243 @@
+    def delete(self, name):
+        path, filename = os.path.split(name)
+        dir_node = self._root.resolve(path)
+        if dir_node is None:
```
Coverage says:
"line 211 didn't jump to line 212, because the condition on line 211 was never true"

pauloxnet commented on django/core/files/storage/memory.py:
```diff
@@ -0,0 +1,243 @@
+        if self.base_url is None:
+            raise ValueError("This file is not accessible via a URL.")
+        url = filepath_to_uri(name)
+        if url is not None:
```
Coverage says:
"line 229 didn't jump to line 231, because the condition on line 229 was never false"

pauloxnet reviewed (approved):
I found the time to do a further review and everything seems to be fine, I just noticed that there are three points of the module not fully covered by the tests. Could you check them?

panicofr commented:
Sure, thanks 👍

felixxm reviewed (commented):
@panicofr Thanks for updates :+1: I tried to run Django test suite with:
```
DEFAULT_FILE_STORAGE = "django.core.files.storage.InMemoryStorage"
```
and some tests don't work:
```
ERROR: test_filename_case_preservation (file_uploads.tests.FileUploadTests)
ERROR: test_override_file_upload_directory_permissions (test_utils.tests.OverrideSettingsTests)
ERROR: test_override_file_upload_permissions (test_utils.tests.OverrideSettingsTests)
ERROR: test_override_media_root (test_utils.tests.OverrideSettingsTests)
ERROR: test_override_media_url (test_utils.tests.OverrideSettingsTests)
FAIL: test_unicode_file_name (file_uploads.tests.FileUploadTests)
FAIL: test_unicode_file_name_rfc2231 (file_uploads.tests.FileUploadTests)
FAIL: test_unicode_file_name_rfc2231_with_double_quotes (file_uploads.tests.FileUploadTests)
FAIL: test_unicode_name_rfc2231 (file_uploads.tests.FileUploadTests)
FAIL: test_unicode_name_rfc2231_with_double_quotes (file_uploads.tests.FileUploadTests)
FAIL: test_media_root_pathlib (model_fields.test_filefield.FileFieldTests)
FAIL: test_move_temporary_file (model_fields.test_filefield.FileFieldTests)
FAIL: test_location_empty (staticfiles_tests.test_finders.TestMiscFinder)
FAIL: test_all_files_more_verbose (staticfiles_tests.test_management.TestFindStatic)
FAIL: test_not_a_directory (file_uploads.tests.DirectoryCreationTests)
FAIL: test_readonly_root (file_uploads.tests.DirectoryCreationTests)

Ran 16153 tests in 175.706s
FAILED (failures=11, errors=5, skipped=1219, expected failures=5)
```
mainly because `InMemoryStorage` lacks some attributes:
```
AttributeError: 'InMemoryStorage' object has no attribute 'directory_permissions_mode'
AttributeError: 'InMemoryStorage' object has no attribute 'file_permissions_mode'
AttributeError: 'InMemoryStorage' object has no attribute 'base_location'
```

We should also credit authors of `dj-inmemorystorage` in the `django/core/files/storage/memory.py` docstring.

panicofr commented:
I fixed almost all the failing tests .

Some of them tests storage behaviour mixing direct filesystem operations (`os.chmod()`, `os.path.exists()`, etc) with calls to storage methods. I refactored them when possible trying to use only storage methods, but I'm not sure it makes sense.

I was thinking instead to skip tests that make use of `os.chmod()` (still failing) when storage class is set to InMemoryStorage (I don't like it, but I can't figured out a better solution).

Some of the involved tests are:
-  test_readonly_root (file_uploads.tests.DirectoryCreationTests)
- test_not_a_directory (file_uploads.tests.DirectoryCreationTests)
- test_move_temporary_file (model_fields.tests.FileFieldTests)
- test_media_root_pathlib (model_fields.tests.FileFieldTests)

What do you think about?

Thanks and merry Xmas everyone :)

pauloxnet commented:
> I fixed almost all the failing tests.
> ... 
> What do you think about?

I don't have specific opinion on using `os` or `pathlib` module in tests.

I tried to execute `coverage` in your branch as specified in the [documention](https://docs.djangoproject.com/en/stable/internals/contributing/writing-code/unit-tests/#code-coverage) and it show some missing line in tests fot the inmemory storage .

```
Name                                Stmts Miss Branch BrPart Cover Missing
-----------------------------------------------------------------------------------------------
django/core/files/storage/memory.py 165   6    48     2      96%   158, 207, 217, 223, 247, 253
```

> Thanks and merry Xmas everyone :)

Marry Xmas to you :)

felixxm commented on tests/model_fields/test_filefield.py:
```diff
@@ -99,9 +99,10 @@ def test_move_temporary_file(self):
             self.assertFalse(
-                os.path.exists(tmp_file_path), "Temporary file still exists"
+                document.myfile.storage.exists(tmp_file_path),
+                "Temporary file still exists",
```
In this case we should leave `os.path.exists()` as this test checks that the temporary uploaded file is moved rather than copied to the destination. We should probably remove a temporary file in `InMemoryStorage._save()`, to get the same behavior with `InMemoryStorage` :thinking:

felixxm commented:
Merged tests updates in c179ad9fe7e82dcb80261aa016f2fe18c8fcc181.

panicofr commented:
@felixxm @pauloxnet I created related [PR 16426](https://github.com/django/django/pull/16426) to add `chmod` method to `FileSystemStorage`. If merged, I could implement the same method for `InMemoryStorage` and use it to fix tests involving operations on file/directory permissions.

What do you think about?

Thanks 😊

felixxm commented:
> What do you think about?

I'm not sure if it's worth doing. I'd skip these tests unless `DEFAULT_FILE_STORAGE` is `FileSystemStorage`.

I can see only 2 failures after reverting changes to the `test_move_temporary_file` (see https://github.com/django/django/pull/16262#discussion_r1059268236):

- `file_uploads.tests.DirectoryCreationTests.test_readonly_root` (I'd skip it):

  ```
  ======================================================================
  FAIL: test_readonly_root (file_uploads.tests.DirectoryCreationTests)
  Permission errors are not swallowed
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/django/tests/file_uploads/tests.py", line 811, in test_readonly_root
      with self.assertRaises(PermissionError):
  AssertionError: PermissionError not raised
  ```
- `file_uploads.tests.DirectoryCreationTests.test_not_a_directory`
  ```
  ======================================================================
  FAIL: test_not_a_directory (file_uploads.tests.DirectoryCreationTests)
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/django/tests/file_uploads/tests.py", line 821, in test_not_a_directory
      with self.assertRaisesMessage(FileExistsError, msg):
    File "/usr/local/lib/python3.10/contextlib.py", line 142, in __exit__
      next(self.gen)
    File "/django/django/test/testcases.py", line 887, in _assert_raises_or_warns_cm
      with func(expected_exception) as cm:
  AssertionError: FileExistsError not raised
  ```
  We could change it to create a file with the default storage API:
  ```diff
  diff --git a/tests/file_uploads/tests.py b/tests/file_uploads/tests.py
  index ecbee48160..14082bbce3 100644
  --- a/tests/file_uploads/tests.py
  +++ b/tests/file_uploads/tests.py
  @@ -12,6 +12,7 @@ from urllib.parse import quote
   from django.core.exceptions import SuspiciousFileOperation
   from django.core.files import temp as tempfile
   from django.core.files.uploadedfile import SimpleUploadedFile, UploadedFile
  +from django.core.files.storage import default_storage
   from django.http.multipartparser import (
       FILE,
       MultiPartParser,
  @@ -815,8 +816,9 @@ class DirectoryCreationTests(SimpleTestCase):
 
       def test_not_a_directory(self):
           # Create a file with the upload directory name
  -        open(UPLOAD_TO, "wb").close()
  -        self.addCleanup(os.remove, UPLOAD_TO)
  +        with SimpleUploadedFile(UPLOAD_TO, b"x") as file:
  +            default_storage.save(UPLOAD_TO, file)
  +        self.addCleanup(default_storage.delete, UPLOAD_TO)
           msg = "%s exists and is not a directory." % UPLOAD_TO
           with self.assertRaisesMessage(FileExistsError, msg):
               with SimpleUploadedFile("foo.txt", b"x") as file:
  ```
  However this raises `SuspiciousFileOperation` with `InMemoryStorage` and shows that in `InMemoryStorage._save()`  should always return `name` relative to the storage root.

felixxm commented on tests/file_storage/tests.py:
```diff
@@ -1157,3 +1160,288 @@ def test_urllib_request_urlopen(self):
+        self.storage.save("file.dat", ContentFile(b"test"))
+        self.assertEqual(self.storage.open("file.dat", "rb").read(), b"test")
+
+    @unittest.skipIf(
+        sys.platform == "win32", "Windows doesn't support moving open files."
+    )
+    def test_removing_temporary_file_after_save(self):
+        """
+        A temporary file should be removed when saved into storage
+        """
+        with TemporaryUploadedFile("test", "text/plain", 1, "utf8") as file:
+            self.storage.save("test.txt", file)
+            self.assertFalse(os.path.exists(file.temporary_file_path()))
```
As far as I'm aware this is already covered by `test_move_temporary_file`. I think we can remove it :thinking:

felixxm commented on tests/model_fields/test_filefield.py:
```diff
@@ -101,7 +101,8 @@ def test_move_temporary_file(self):
             Document.objects.create(myfile=tmp_file)
             self.assertFalse(
-                os.path.exists(tmp_file_path), "Temporary file still exists"
+                os.path.exists(tmp_file_path),
+                "Temporary file still exists",
```
```suggestion
                os.path.exists(tmp_file_path), "Temporary file still exists"
```

felixxm commented on tests/file_uploads/tests.py:
```diff
@@ -804,6 +806,10 @@ def setUp(self):
     @unittest.skipIf(
         sys.platform == "win32", "Python on Windows doesn't have working os.chmod()."
     )
+    @unittest.skipIf(
+        settings.DEFAULT_FILE_STORAGE != "django.core.files.storage.FileSystemStorage",
+        "Only FileSystemStorage supports file and directory permissions",
+    )
```
Maybe:
```suggestion
    @override_settings(DEFAULT_FILE_STORAGE="django.core.files.storage.FileSystemStorage")
```

felixxm commented:
@panicofr Thanks :+1:

I pushed final edits.

pauloxnet commented:
LGTM

carltongibson commented on docs/ref/files/storage.txt:
```diff
@@ -74,6 +74,39 @@ The ``FileSystemStorage`` class
+
+    The :class:`~django.core.files.storage.InMemoryStorage` class implements
+    a memory-based file storage. It has no persistence, but can be useful for
+    speeding tests up by avoiding disk access.
```
Really minor by I think I'd say... 

```suggestion
    speeding up tests by avoiding disk access.
```

carltongibson commented on docs/releases/4.2.txt:
```diff
@@ -85,6 +85,12 @@ The Breach (HTB) paper`_.
+----------------------
+
+The new ``django.core.files.storage.InMemoryStorage`` class provides a
+non-persistent storage useful for speeding tests up by avoiding disk access.
```
```suggestion
non-persistent storage useful for speeding up tests by avoiding disk access.
```

felixxm replied:
Updated :+1:
